src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala

```diff
@@ -0,0 +1,325 @@
+    ProjectedDataModelInstance(
+      externalId = dmi.properties
+        .flatMap(_.get("externalId"))
+        .flatMap(_.asString)
+        .getOrElse(
+          throw new CdfSparkException("Can't read data model instance, `externalId` is missing.")),
+      properties = requiredPropsArray.map { name: String =>
+        val value = dmiProperties.getOrElse(name, Json.Null)
+        parseFromJson(value, propertyTypes(name)._1)
+          .getOrElse(throw new CdfSparkException(s"Unexpected value $value in property $name"))
+
+      }.toArray
+    )
+  }
+
+  def getStreams(filters: Array[Filter], requiredColumns: Array[String])(
+      client: GenericClient[IO],
+      limit: Option[Int],
+      numPartitions: Int): Seq[Stream[IO, ProjectedDataModelInstance]] = {
+    val requiredPropsArray: Seq[String] = if (requiredColumns.isEmpty) {
+      modelPropertyStructFields.map(_.name)
+    } else {
+      requiredColumns
+    }
+
+    // TODO: Implement pushdown filters
+    val dmiQuery = DataModelInstanceQuery(
+      modelExternalId = modelExternalId,
+      filter = None,
+      sort = None,
+      limit = limit,
+      cursor = None)
+    Seq(
+      alphaClient.dataModelInstances
+        .queryStream(dmiQuery, limit)
+        .map(r => toProjectedInstance(r, requiredPropsArray)))
+  }
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] =
+    SdkV1Rdd[ProjectedDataModelInstance, String](
+      sqlContext.sparkContext,
+      config,
+      (item: ProjectedDataModelInstance, _) => toRow(item),
+      uniqueId,
+      getStreams(filters, requiredColumns)
+    )
+
+  def insert(rows: Seq[Row]): IO[Unit] =
+    throw new CdfSparkException("Insert is not supported for data model instances. Use upsert instead.")
+
+  def update(rows: Seq[Row]): IO[Unit] =
+    throw new CdfSparkException("Update is not supported for data model instances. Use upsert instead.")
+
+  override def delete(rows: Seq[Row]): IO[Unit] =
+    throw new CdfSparkException("Delete is not supported for data model instances.")
+
+  def fromRow(schema: StructType): Row => DataModelInstance = {
+    val externalIdIndex = schema.fieldNames.indexOf("externalId")
+    if (externalIdIndex < 0) {
+      throw new CdfSparkException("Can't upsert data model instances, `externalId` is missing.")
+    }
+
+    val properties: Array[(Int, String, (String, Boolean))] = schema.fields.zipWithIndex
+      .map {
+        case (field: StructField, index: Int) =>
+          val propertyType = propertyTypes.getOrElse(
+            field.name,
+            throw new CdfSparkException(
+              s"Can't insert property `${field.name}` " +
+                s"into data model $modelExternalId, the property does not exist in the definition")
+          )
+          (index, field.name, propertyType)
+      }
+
+    def parseRow(row: Row): DataModelInstance = {
+      val externalId = row.get(externalIdIndex) match {
```

Comment:
Looks like this value isn't being used?

Reply:
it is just to validate i can make it `val _` instead of `val externalId`